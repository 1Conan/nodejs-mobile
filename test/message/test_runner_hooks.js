// Flags: --no-warnings
'use strict';
require('../common');
const assert = require('assert');
const { test, describe, it, before, after, beforeEach, afterEach } = require('node:test');

describe('describe hooks', () => {
  const testArr = [];
  before(function() {
    testArr.push('before ' + this.name);
  });
  after(function() {
    testArr.push('after ' + this.name);
    assert.deepStrictEqual(testArr, [
      'before describe hooks',
      'beforeEach 1', '1', 'afterEach 1',
      'beforeEach 2', '2', 'afterEach 2',
      'before nested',
      'beforeEach nested 1', 'nested 1', 'afterEach nested 1',
      'beforeEach nested 2', 'nested 2', 'afterEach nested 2',
      'after nested',
      'after describe hooks',
    ]);
  });
  beforeEach(function() {
    testArr.push('beforeEach ' + this.name);
  });
  afterEach(function() {
    testArr.push('afterEach ' + this.name);
  });

  it('1', () => testArr.push('1'));
  it('2', () => testArr.push('2'));

  describe('nested', () => {
    before(function() {
      testArr.push('before ' + this.name);
    });
    after(function() {
      testArr.push('after ' + this.name);
    });
    beforeEach(function() {
      testArr.push('beforeEach ' + this.name);
    });
    afterEach(function() {
      testArr.push('afterEach ' + this.name);
    });
    it('nested 1', () => testArr.push('nested 1'));
    it('nested 2', () => testArr.push('nested 2'));
  });
});

describe('before throws', () => {
  before(() => { throw new Error('before'); });
  it('1', () => {});
  it('2', () => {});
});

describe('after throws', () => {
  after(() => { throw new Error('after'); });
  it('1', () => {});
  it('2', () => {});
});

describe('beforeEach throws', () => {
  beforeEach(() => { throw new Error('beforeEach'); });
  it('1', () => {});
  it('2', () => {});
});

describe('afterEach throws', () => {
  afterEach(() => { throw new Error('afterEach'); });
  it('1', () => {});
  it('2', () => {});
});

test('test hooks', async (t) => {
  const testArr = [];
  t.beforeEach((t) => testArr.push('beforeEach ' + t.name));
  t.afterEach((t) => testArr.push('afterEach ' + t.name));
  await t.test('1', () => testArr.push('1'));
  await t.test('2', () => testArr.push('2'));

  await t.test('nested', async (t) => {
    t.beforeEach((t) => testArr.push('nested beforeEach ' + t.name));
    t.afterEach((t) => testArr.push('nested afterEach ' + t.name));
    await t.test('nested 1', () => testArr.push('nested1'));
    await t.test('nested 2', () => testArr.push('nested 2'));
  });

  assert.deepStrictEqual(testArr, [
    'beforeEach 1', '1', 'afterEach 1',
    'beforeEach 2', '2', 'afterEach 2',
    'beforeEach nested',
    'nested beforeEach nested 1', 'nested1', 'nested afterEach nested 1',
    'nested beforeEach nested 2', 'nested 2', 'nested afterEach nested 2',
    'afterEach nested',
  ]);
});

test('t.beforeEach throws', async (t) => {
  t.beforeEach(() => { throw new Error('beforeEach'); });
  await t.test('1', () => {});
  await t.test('2', () => {});
});

test('t.afterEach throws', async (t) => {
  t.afterEach(() => { throw new Error('afterEach'); });
  await t.test('1', () => {});
  await t.test('2', () => {});
});
