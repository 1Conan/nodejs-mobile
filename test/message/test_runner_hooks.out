TAP version 13
# Subtest: describe hooks
    # Subtest: 1
    ok 1 - 1
      ---
      duration_ms: *
      ...
    # Subtest: 2
    ok 2 - 2
      ---
      duration_ms: *
      ...
    # Subtest: nested
        # Subtest: nested 1
        ok 1 - nested 1
          ---
          duration_ms: *
          ...
        # Subtest: nested 2
        ok 2 - nested 2
          ---
          duration_ms: *
          ...
        1..2
    ok 3 - nested
      ---
      duration_ms: *
      ...
    1..3
ok 1 - describe hooks
  ---
  duration_ms: *
  ...
# Subtest: before throws
    # Subtest: 1
    not ok 1 - 1
      ---
      duration_ms: *
      failureType: 'cancelledByParent'
      error: 'test did not finish before its parent and was cancelled'
      code: 'ERR_TEST_FAILURE'
      ...
    # Subtest: 2
    not ok 2 - 2
      ---
      duration_ms: *
      failureType: 'cancelledByParent'
      error: 'test did not finish before its parent and was cancelled'
      code: 'ERR_TEST_FAILURE'
      ...
    1..2
not ok 2 - before throws
  ---
  duration_ms: *
  failureType: 'hookFailed'
  error: 'failed running before hook'
  code: 'ERR_TEST_FAILURE'
  stack: |-
    *
    *
    *
    *
    *
    *
    *
    *
    *
  ...
# Subtest: after throws
    # Subtest: 1
    ok 1 - 1
      ---
      duration_ms: *
      ...
    # Subtest: 2
    ok 2 - 2
      ---
      duration_ms: *
      ...
    1..2
not ok 3 - after throws
  ---
  duration_ms: *
  failureType: 'hookFailed'
  error: 'failed running after hook'
  code: 'ERR_TEST_FAILURE'
  stack: |-
    *
    *
    *
    *
    *
    *
    *
    *
    *
  ...
# Subtest: beforeEach throws
    # Subtest: 1
    not ok 1 - 1
      ---
      duration_ms: *
      failureType: 'hookFailed'
      error: 'failed running beforeEach hook'
      code: 'ERR_TEST_FAILURE'
      stack: |-
        *
        *
        *
        *
        *
        *
        *
        *
        *
        *
      ...
    # Subtest: 2
    not ok 2 - 2
      ---
      duration_ms: *
      failureType: 'hookFailed'
      error: 'failed running beforeEach hook'
      code: 'ERR_TEST_FAILURE'
      stack: |-
        *
        *
        *
        *
        *
        *
        *
        *
        *
        *
      ...
    1..2
not ok 4 - beforeEach throws
  ---
  duration_ms: *
  failureType: 'subtestsFailed'
  error: '2 subtests failed'
  code: 'ERR_TEST_FAILURE'
  ...
# Subtest: afterEach throws
    # Subtest: 1
    not ok 1 - 1
      ---
      duration_ms: *
      failureType: 'hookFailed'
      error: 'failed running afterEach hook'
      code: 'ERR_TEST_FAILURE'
      stack: |-
        *
        *
        *
        *
        *
        *
        *
        *
        *
        *
      ...
    # Subtest: 2
    not ok 2 - 2
      ---
      duration_ms: *
      failureType: 'hookFailed'
      error: 'failed running afterEach hook'
      code: 'ERR_TEST_FAILURE'
      stack: |-
        *
        *
        *
        *
        *
        *
        *
        *
        *
        *
      ...
    1..2
not ok 5 - afterEach throws
  ---
  duration_ms: *
  failureType: 'subtestsFailed'
  error: '2 subtests failed'
  code: 'ERR_TEST_FAILURE'
  ...
# Subtest: test hooks
    # Subtest: 1
    ok 1 - 1
      ---
      duration_ms: *
      ...
    # Subtest: 2
    ok 2 - 2
      ---
      duration_ms: *
      ...
    # Subtest: nested
        # Subtest: nested 1
        ok 1 - nested 1
          ---
          duration_ms: *
          ...
        # Subtest: nested 2
        ok 2 - nested 2
          ---
          duration_ms: *
          ...
        1..2
    ok 3 - nested
      ---
      duration_ms: *
      ...
    1..3
ok 6 - test hooks
  ---
  duration_ms: *
  ...
# Subtest: t.beforeEach throws
    # Subtest: 1
    not ok 1 - 1
      ---
      duration_ms: *
      failureType: 'hookFailed'
      error: 'failed running beforeEach hook'
      code: 'ERR_TEST_FAILURE'
      stack: |-
        *
        *
        *
        *
        *
        *
        *
        *
        *
        *
      ...
    # Subtest: 2
    not ok 2 - 2
      ---
      duration_ms: *
      failureType: 'hookFailed'
      error: 'failed running beforeEach hook'
      code: 'ERR_TEST_FAILURE'
      stack: |-
        *
        *
        *
        *
        *
        *
        *
        *
        *
        *
      ...
    1..2
not ok 7 - t.beforeEach throws
  ---
  duration_ms: *
  failureType: 'subtestsFailed'
  error: '2 subtests failed'
  code: 'ERR_TEST_FAILURE'
  ...
# Subtest: t.afterEach throws
    # Subtest: 1
    not ok 1 - 1
      ---
      duration_ms: *
      failureType: 'hookFailed'
      error: 'failed running afterEach hook'
      code: 'ERR_TEST_FAILURE'
      stack: |-
        *
        *
        *
        *
        *
        *
        *
        *
        *
        *
      ...
    # Subtest: 2
    not ok 2 - 2
      ---
      duration_ms: *
      failureType: 'hookFailed'
      error: 'failed running afterEach hook'
      code: 'ERR_TEST_FAILURE'
      stack: |-
        *
        *
        *
        *
        *
        *
        *
        *
        *
        *
      ...
    1..2
not ok 8 - t.afterEach throws
  ---
  duration_ms: *
  failureType: 'subtestsFailed'
  error: '2 subtests failed'
  code: 'ERR_TEST_FAILURE'
  ...
1..8
# tests 8
# pass 2
# fail 6
# cancelled 0
# skipped 0
# todo 0
# duration_ms *
